sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler.asciidoc

```diff
@@ -0,0 +1,501 @@
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_KERNEL_COMPILER`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New kernel bundle state
+
+This extension adds the `ext_oneapi_source` enumerator to `sycl::bundle_state`
+to identify a kernel bundle that is represented as a source code string.
+
+```
+namespace sycl {
+
+enum class bundle_state : /*unspecified*/ {
+  // ...
+  ext_oneapi_source
+};
+
+} // namespace sycl
+```
+
+=== New enumerator of kernel source languages
+
+This extension adds the `source_language` enumeration, which identifies
+possible languages for a kernel bundle that is in `ext_oneapi_source` state:
+
+```
+namespace sycl::ext::oneapi::experimental {
+
+enum class source_language : /*unspecified*/ {
+  // see below
+};
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+However, there are no enumerators defined by this extension.
+Instead, this enumeration is an extension point for other extensions, which can
+specify the exact semantics of each possible kernel language.
+
+=== New free functions to create and build kernel bundles
+
+This extension adds the following new free functions to create and build a
+kernel bundle in `ext_oneapi_source` state.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+kernel_bundle<bundle_state::ext_oneapi_source> create_kernel_bundle_from_source(
+  const context& ctxt,
+  source_language lang,
+  const std::string& source)
+
+} // namespace sycl::ext::oneapi::experimental {
+----
+!====
+
+_Effects:_ Creates a new kernel bundle that represents a kernel that is defined
+by the source code string `source` using the language `lang`.
+The bundle is associated with the context `ctxt`, and kernels from this bundle
+may only be submitted to a queue that shares the same context.
+The bundle's set of associated devices is the set of devices contained in
+`ctxt`.
+
+_Returns:_ The newly created kernel bundle, which has `ext_oneapi_source`
+state.
+
+_Throws:_
+
+* An `exception` with the `errc::invalid` error code if the source language
+  `lang` is not supported by the backend associated with `ctxt`.
+
+_Remarks:_ Calling this function does not attempt to compile the source code.
+As a result, syntactic errors in the source code string are not diagnosed by
+this function.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+bool is_source_kernel_bundle_supported(backend be, source_language lang)
```

Comment:
What is the use-case for this?  `create_kernel_bundle_from_source` already throws an exception if this would return `false`, and a developer can't change the backend associated with the context they'd like to build against.

If the intent is to allow developers to use this to filter platforms (i.e., to make sure they select a platform where they can compile their source code), should it operate on a `platform` instead?  I don't think we should expose the concept of a `backend` unless we really need to.

If you agree, there are a few other places we could replace "backend" with "platform" in the descriptions of function behavior.

Reply:
This is the query function that applications can use to determine if `create_kernel_bundle_from_source` will succeed.  I debated whether the parameter should be `backend` or `platform`.  I chose backend over platform because I thought that was the true dependency.  If an implementation exposes several platforms over the same backend, it seemed wasteful to require the application to test them all.

What is the argument for preferring platform?  Both platform and backend are existing SYCL concepts that are exposed to the user.

Reply:
My argument is that `platform` is a higher-level concept, and a developer is more likely to be working with platforms than backends.  A user can iterate over all the platforms and select one (with `platform::get_platforms()`), but they can't iterate over backends.

I was trying to think of how I might use this, and it seemed like I'd always have to call this as `is_source_kernel_bundle_supported(platform.get_backend(), lang)`, so it'd be easier to accept a `platform` directly.

As I type this, I realize that device selection is probably even more important than platform selection.  If a backend has support for a language, is that actually even sufficient to choose a device?  Could there be cases where a vendor decides to put multiple hardware types in the same platform, but can only compile some languages for some of the devices?

If so, `platform::can_compile(lang, device)` might be closer to what we need.  You could use that in a device selector too, as `device.get_platform().can_compile(lang, device)`.

Reply:
You've convinced me that the query shouldn't be for the backend.  Rather than adding something like `platform::can_compile(lang, device)`, wouldn't it make more sense to add `device::can_compile(lang)`?

Reply:
Good point.  Yes, that's even simpler.  A device can only belong to one platform, so `device::can_compile(lang)` makes sense.

Reply:
Done in 10993279ac84ab6da6a1c81f78808050bc58575a